### DISCUSSION TITLE
How does PNG transform into KTX2

### DISCUSSION THREAD

**Discussion by baiwenbo1997** [KhronosGroup/KTX-Software#559] [Q&A] [answered]
I am an initiator. I would like to ask how the PNG can be successfully converted to KTX2. I refer to https://github.khronos.org/ktx-software/libktx/index.html#writektx

**Comment by MarkCallow** [comment, marked as answer]
@baiwenbo1997 the `ktxTexture2_Write*` functions write a ktxTexture2 object in memory to a file. That is they serialize existing KTX2 data. They do not do any conversion. Use `toktx` to create KTX2 files from PNG files. If you want to write your own software to do this, also look at `toktx` to see how to do it. The steps for a color texture are (some steps are different if handling normal maps)

1. Read PNG header
2. Select matching Vulkan format
3. Create ktxTexture2 object with that format
4. Read PNG data expanding bit-width if necessary
5. Add data to base level of the ktxTexture2 object
6. Generate mip levels or read an additional file per level, if desired, and add to ktxTexture2 object.
7. Write the ktxTexture2 object to a file

**Reply by baiwenbo1997** [reply]
Thank you for your reply, I can currently write the PNG file into KTX2 file, but use http://texture.superzeng.top/ The tool is unable to view, then it is a problem that KTX2 files are written, but I don't know where is it? something is wrong
![image](https://user-images.githubusercontent.com/44999478/163969057-a081c347-ad25-417e-9ede-af1afe9b9682.png)

**Reply by MarkCallow** [reply]
I do not understand the problem you are having and I do not know what http://texture.superzeng.top/ does so I can't help. There are a couple of things wrong or potentially wrong with the code extract you show:

1. A `ktxTexture2_Write*` function must be called after `ktxTexture_SetImageFromMemory` to serialize the data to a file. If your issue is that you can't find the output file in your file system this is likely the cause.
2. You cannot hardcode VK_FORMAT_R8G8B8_UNORM as the PNG file may have an alpha channel or may have 16-bits per channel.

If your issue is that http://texture.superzeng.top/ is not displaying the file, you can use `ktx2check` to check if the file is valid KTX2. If it is, then complain to the web site.

**Reply by baiwenbo1997** [reply]
Sorry, I may not be clear enough before me. The problem I faced is that it can successfully generate a KTX2 file. The reason is that the PNG format decoding error causes an invalid KTX2 file, so I want to teach how do you correctly decode and output a valid KTX2 file for the PNG file?

**Reply by MarkCallow** [reply]
You have to unpack and decompress the image in the PNG file and pass a pointer to that image to ktxTexture_SetImageFromMemory.

Look at the source code for [`toktx`](../tree/master/tools/toktx) to see an example. `pngimage.cc` determines the image format and unpacks and decompresses the image using `lodepng.cc`. `toktx.cc` creates the ktxTexture2, uploads the image to it and writes the file. There is a lot of code because it handles most PNG formats, does color management and handles either a PNG file per mip level or generation of mip levels from a single input image. Therefore it is not easy to understand.

**Reply by baiwenbo1997** [reply]
I still have a question here. Can I convert it to a PNG format if it is a KTX2 format file? If you need to decompress the KTX2 file first, switch to PNG after decompression

**Reply by MarkCallow** [reply]
It is possible. Yes you have to decompress the KTX2 file if it is compressed. In any case you need to extract the image(s) from the KTX2 file. You may have to convert to multiple PNG files if the KTX2 file contains multiple images.

There is no tool in the KTX-Software suite to do this at present. the [Basis Universal](https://github.com/BinomialLLC/basis_universal) command line compression tool `basisu` can create PNG files from KTX2 files whose payload is one of the Basis Universal formats (ETC1S/LZ or UASTC). I think you can also export PNGs from [PVRTexTool](https://developer.imaginationtech.com/pvrtextool/) for some other payload formats. (I have not tried it recently so do not know how well it supports KTX2 and I also do not know if they have yet provided support for BC3 (S3TC) formats on macOS. They were previously afraid of Creative Labs lawyers but the patent lapsed several years ago now.)

**Comment by baiwenbo1997** [comment]
Thank you very much, I have read the PNG format to RGBA8 memory, I want to convert this memory into KTX2 file, the interface that should be used is ktxtexture_setimageFromMemory (), so I need to read the read PNG memory how to convert Can join the image to join the image

------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "KhronosGroup/KTX-Software"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年4月17日(星期天) 下午5:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [KhronosGroup/KTX-Software] How does PNG transform into KTX2 (Discussion #559)





 
@baiwenbo1997 the ktxTexture2_Write* functions write a ktxTexture2 object in memory to a file. That is they serialize existing KTX2 data. They do not do any conversion. Use toktx to create KTX2 files from PNG files. If you want to write your own software to do this, also look at toktx to see how to do it. The steps for a color texture are (some steps are different if handling normal maps)
  
Read PNG header
 
Select matching Vulkan format
 
Create ktxTexture2 object with that format
 
Read PNG data expanding bit-width if necessary
 
Add data to base level of the ktxTexture2 object
 
Generate mip levels or read an additional file per level, if desired, and add to ktxTexture2 object.
 
Write the ktxTexture2 object to a file

**Reply by MarkCallow** [reply]
If you have read the PNG format to RGBA8 memory then no further conversion is necessary. After setting the image in the ktxTexture[12] object with `ktxTexture_SetImageFromMemory you can use ktxTexture[12]_WriteToFile to write the KTX2 file. If you want to compress it with BasisU or ASTC, call the appropriate compress function before writing the file.

I don't understand what you are asking about joining images.
